aphrodite/modeling/layers/sampler.py

```diff
@@ -318,12 +318,18 @@ def _apply_temperatures(
     normalized_entropies = dynatemp_entropies.div_(dynatemp_max_entropies)
     dyn_temp = (dynatemp_mins + (dynatemp_maxs - dynatemp_mins) *
                 normalized_entropies.pow_(dynatemp_exps))
-
     temperatures[dynatemp_mask] = dyn_temp
-    temperatures[temperatures <= 0.0] = 1.0
-    # Use float32 to apply temp.
-    # Use in-place division to avoid creating a new tensor.
-    logits = logits.to(torch.float)
+  
+    # There isn't a "safe" temperature range for fp16 logits.
+    # This value was chosen because 1/2e-5 is just under the 65k fp16 max,
+    # meaning that this temperature well-uses the fp16 space after the offset.
+    temperatures[temperatures <= 2e-5] = 2e-5
+  
+    # To prevent saturation of top logits, we shift the range to [-inf, 1]
+    # Why align to 1, instead of 0? Because [0, 1] holds 25% of all floats.
+    # Why mask? So we aren't potentially discarding data in milder temps.
+    low_temps = temperatures < 0.1
+    logits[low_temps] -= logits.max(dim=-1, keepdim=True).values[low_temps] - 1
```

Comment:
Any performance impacts here? Considering the range shifting adds an extra max op and indexing for low temps. I can run a benchmark later if you can't

Reply:
No idea, but historically I've never seen sampling take up more than a percent or two of realistic workloads.
Also, keep in mind, almost every other sampler does a full _sort_ on the logits, let alone a mere O(N) `max`.